src/xcode/ENA/ENA/Source/Scenes/QRScanner/QRScannerCoordinator.swift

```diff
@@ -332,5 +375,123 @@ class QRScannerCoordinator {
 
 		presentationController.present(alert, animated: true)
 	}
+
+	private func showTestRestoredFromBinAlert(
+		recycleBinItem: RecycleBinItem
+	) {
+		let alert = UIAlertController(
+			title: AppStrings.UniversalQRScanner.testRestoredFromBinAlertTitle,
+			message: AppStrings.UniversalQRScanner.testRestoredFromBinAlertMessage,
+			preferredStyle: .alert
+		)
+		alert.addAction(
+			UIAlertAction(
+				title: AppStrings.Common.alertActionOk,
+				style: .default,
+				handler: { [weak self] _ in
+					guard let self = self else { return }
+
+					switch self.recycleBin.canRestore(recycleBinItem) {
+					case .success:
+						self.qrScannerViewController?.dismiss(animated: true) {
+							switch self.presenter {
+							case .submissionFlow, .onBehalfFlow:
+								let parentPresentingViewController = self.parentViewController?.presentingViewController
+
+								// Dismiss submission/on behalf submission flow
+								parentPresentingViewController?.dismiss(animated: true) {
+									self.parentViewController = parentPresentingViewController
+									self.restoreAndShow(recycleBinItem: recycleBinItem)
+								}
+							case .checkinTab, .certificateTab, .universalScanner:
+								self.restoreAndShow(recycleBinItem: recycleBinItem)
+							case .none:
+								break
+							}
+						}
+					case .failure(.testError(.testTypeAlreadyRegistered)):
+						self.showTestOverwriteNotice(recycleBinItem: recycleBinItem)
+					}
+				}
+			)
+		)
+
+		qrScannerViewController?.present(alert, animated: true)
+	}
+
+	private func showTestOverwriteNotice(
+		recycleBinItem: RecycleBinItem
+	) {
+		guard case let .coronaTest(coronaTest) = recycleBinItem.item else {
+			return
+		}
+
+		let footerViewModel = FooterViewModel(
+			primaryButtonName: AppStrings.ExposureSubmission.OverwriteNotice.primaryButton,
+			isSecondaryButtonHidden: true
+		)
+
+		let overwriteNoticeViewController = TestOverwriteNoticeViewController(
+			testType: coronaTest.type,
+			didTapPrimaryButton: { [weak self] in
+				self?.restoreAndShow(recycleBinItem: recycleBinItem)
+			},
+			didTapCloseButton: { [weak self] in
+				self?.parentViewController?.dismiss(animated: true)
+			}
+		)
+
+		let footerViewController = FooterViewController(footerViewModel)
+		let topBottomViewController = TopBottomContainerViewController(
+			topController: overwriteNoticeViewController,
+			bottomController: footerViewController
+		)
+
+		let navigationController = NavigationControllerWithLargeTitle(rootViewController: topBottomViewController)
+
+		switch self.presenter {
```

Comment:
This logic looks similar to other calls. Can we have one func to handle it at one place?

Reply:
The logic looks similar but differs in details, I don't see an easy way here to handle it in a uniform way.